modules/consentManagement.js

```diff
@@ -0,0 +1,249 @@
+/**
+ * This module adds GDPR consentManagement support to prebid.js.  It interacts with
+ * supported CMPs (Consent Management Platforms) to grab the user's consent information
+ * and make it available for any GDPR supported adapters to read/pass this information to
+ * their system.
+ */
+import * as utils from 'src/utils';
+import { config } from 'src/config';
+import { gdprDataHandler } from 'src/adaptermanager';
+import includes from 'core-js/library/fn/array/includes';
+
+const DEFAULT_CMP = 'iab';
+const DEFAULT_CONSENT_TIMEOUT = 10000;
+const DEFAULT_ALLOW_AUCTION_WO_CONSENT = true;
+
+export let userCMP;
+export let consentTimeout;
+export let allowAuction;
+
+let consentData;
+
+let context;
+let args;
+let nextFn;
+
+let timer;
+let haveExited;
+
+// add new CMPs here, with their dedicated lookup function
+const cmpCallMap = {
+  'iab': lookupIabConsent
+};
+
+/**
+ * This function handles interacting with the AppNexus CMP to obtain the consentObject value of the user.
+ * Given the async nature of the CMP's API, we pass in acting success/error callback functions to exit this function
+ * based on the appropriate result.
+ * @param {function(string)} cmpSuccess acts as a success callback when CMP returns a value; pass along consentObject (string) from CMP
+ * @param {function(string)} cmpError acts as an error callback while interacting with CMP; pass along an error message (string)
+ */
+function lookupIabConsent(cmpSuccess, cmpError) {
+  // let callId = 0;
+
+  if (window.__cmp) {
+    window.__cmp('getVendorConsents', null, cmpSuccess);
+  } else {
+    /** START OF STOCK CODE FROM IAB 1.1 CMP SEC */
+    // find the CMP frame
+    let f = window;
+    let cmpFrame;
+    while (!cmpFrame) {
+      try {
+        if (f.frames['__cmpLocator']) cmpFrame = f;
+      } catch (e) {}
+      if (f === window.top) break;
+      f = f.parent;
+    }
+
+    let cmpCallbacks = {};
+
+    /* Setup up a __cmp function to do the postMessage and stash the callback.
+      This function behaves (from the caller's perspective identicially to the in-frame __cmp call */
+    window.__cmp = function(cmd, arg, callback) {
```

Comment:
This makes the assumption that the CMP is bound to the window in which Prebid loads. For safety, we have always loaded Prebid (and our libraries too) inside of a friendly frame, so in our case, we would need this to reference the top-most window.

```typescript
export function requestCMP(): void {
    let topWindow = WindowElement.getTopMostParentWindow();
    let createHandler = () => {
        let commandQueue: any = [];
        let cmp = function(command: any, parameter: any, callback: any){
            commandQueue.push({
                command: command,
                parameter: parameter,
                callback: callback
            });
        };
        (cmp as any).commandQueue = commandQueue;
        (cmp as any).receiveMessage = (event: any) => {
            let data = event && event.data && event.data.__cmp;
            if (data) {
                commandQueue.push({
                    callId: data.callId,
                    command: data.command,
                    parameter: data.parameter,
                    event: event
                });
            }
        };
        (cmp as any).config = {
            "globalConsentLocation": "https://acdn.adnxs.com/cmp/docs/portal.html",
            "storeConsentGlobally": true,
            "storePublisherData": false
        };
        return cmp;
    };
    (topWindow as any).__cmp = createHandler();
    topWindow.addEventListener('message', (event) => {
        (topWindow as any).__cmp.receiveMessage(event);
    });
    (topWindow as any).__cmp('showConsentTool');
    Async.loadJSIn(topWindow, "https://acdn.adnxs.com/cmp/docs/../cmp.bundle.js");
}
```

As the CMP will always be in the top-most window, can we change this?

Reply:
Hi @cwbeck, 

The part of the code you referenced is actually part of the code where we have already determined that the CMP is not in the same window as the prebid code.  So it's not making one direction assumption as to where the CMP exists, as this can vary per publisher's implementation.

To provide some additional context on the function in question:
This function is part of the stock IAB code that's part of the 1.1 CMP spec to use the CMP's iframe locator (ie `__cmpLocator`) to detect the exact frame that the CMP is located.  

The IAB's code works basically in 3 parts:
- locates the proper window/frame that's housing the CMP code (by looking for the parent window that contains the `__cmpLocator` iframe)
- it then writes it's own local version of the `window.__cmp` API to handle the natural API call and encase it in a `postMessage` call to bypass the iframe.
- it finally adds an `addEventListener` to wait for the `message` response from the CMP to grab the consent data and do the next steps.

Please let me know if you have any further questions on this point.